tsl/src/continuous_aggs/invalidation_plugin.c

```diff
@@ -0,0 +1,533 @@
+	 *
+	 * This means that if oldtuple is NULL, it is sufficient to use newtuple
+	 * to decide what range that was updated.
+	 */
+	if (change->data.tp.oldtuple)
+		invalidations_insert_tuple(data, relinfo, relation, as_tuple(change->data.tp.oldtuple));
+	invalidations_insert_tuple(data, relinfo, relation, as_tuple(change->data.tp.newtuple));
+}
+
+static void
+invalidations_record_delete(InvalidationsPluginData *data, TrackRelInfo *relinfo,
+							LogicalDecodingContext *ctx, Relation relation,
+							ReorderBufferChange *change)
+{
+	/*
+	 * We need a replica identity to deal with deletes. If the table does not
+	 * have a replica identity, we would have to invalidate the entire table,
+	 * so instead we throw an error if the delete does not have an oldtuple.
+	 */
+	if (!change->data.tp.oldtuple)
+		ereport(ERROR,
+				errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				errmsg("table is missing replica identity"),
+				errdetail("Table \"%s\" need to have a replica identity to handle DELETE.",
+						  RelationGetRelationName(relation)));
+	invalidations_insert_tuple(data, relinfo, relation, as_tuple(change->data.tp.oldtuple));
+}
+
+static List *
+parse_output_parameters(List *options)
+{
+	ListCell *option;
+	List *relinfo_list = NIL;
+
+	/*
+	 * Options are pairs of relation names and the attribute to read.
+	 *
+	 * We do not check if the "key" is a hypertable since we do not want to
+	 * rely on the TimescaleDB metadata.
+	 */
+	foreach (option, options)
+	{
+		DefElem *elem = lfirst(option);
+		const char *name = elem->defname;
+#if PG16_LT
+		List *qname = stringToQualifiedNameList(name);
+#else
+		List *qname = stringToQualifiedNameList(name, NULL);
+#endif
+		TrackRelInfo *relinfo = palloc0(sizeof(TrackRelInfo));
+		RangeVar *hypertable_rv = makeRangeVarFromNameList(qname);
+		relinfo->hypertable_relid = RangeVarGetRelid(hypertable_rv, AccessShareLock, true);
+		if (!OidIsValid(relinfo->hypertable_relid))
+			ereport(ERROR,
+					errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("relation \"%s\" does not exist", name));
+
+		NameData attname;
+		namestrcpy(&attname, strVal(elem->arg));
+
+		relinfo->attnum = get_attnum(relinfo->hypertable_relid, NameStr(attname));
+		if (relinfo->attnum == InvalidAttrNumber)
+			ereport(ERROR,
+					errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("attribute \"%s\" does not exist", NameStr(attname)));
+
+		relinfo_list = lappend(relinfo_list, relinfo);
+	}
+	return relinfo_list;
+}
+
+static void
+invalidations_write_record(InvalidationCacheEntry *entry, bool last_write,
+						   InvalidationsContext *args)
+{
+	LogicalDecodingContext *ctx = args->ctx;
+
+	TS_DEBUG_LOG("write entry: hypertable_id=%u, lowest_modified_value=%" PRId64 ", "
+				 "greatest_modified_value=%" PRId64,
+				 entry->hypertable_relid,
+				 entry->lowest_modified,
+				 entry->highest_modified);
+
+	InvalidationEntry inval;
+	inval.version = INVALIDATION_ENTRY_CURRENT_VERSION;
```

Comment:
The encode function also writes this version to this field, so it doesn't seem needed. We should have one way to deal with version, either the decode function does it or the caller decides. 

I think we probably don't need a versioned struct at all because the encode function can just prepend the version in the output stream. See other comments on this, especially in `invalidation_record.h`